Add vk gate support

vladimirshkoda opened the pull request:
Close #4 

Add:
- Function for obtaining authorization url
- Function for posting to group
- Mini Vk API client.

vladimirshkoda committed:
Add vk group support

b0g3r commented on postpost/api/vkontakte.py:
```diff
@@ -0,0 +1,32 @@
+
+def get_authorization_url(client_id: int, api_version: float) -> str:
+    """
+    Returns a string with url for authorization.
```
Стоит дописать, что используется для получения access-token'а, возможно приложить ссылочку или пример использования

vladimirshkoda replied:
Сделано.

b0g3r commented on postpost/api/tasks.py:
```diff
@@ -37,8 +38,30 @@ def send_post_to_telegram_channel(scheduled_post_id: int):
+    Celery task which tries to send post to vk and changes status of platform post.
+    """
+    post = PlatformPost.objects.select_related('publication').get(id=scheduled_post_id)
+    vk_response = vkontakte.send_post_to_group(
```
Я бы здесь сделал более высокоуровнево: просто функция, которая принимает пост. А внутри уже экстрактится текст, аттачменты и прикидывается в низкоуровневую функцию

vladimirshkoda replied:
Это я сделаю в будущем каком-нибудь коммите, потому что сейчас непонятно что пост будет содержать.

b0g3r commented on postpost/api/tasks.py:
```diff
@@ -37,8 +38,30 @@ def send_post_to_telegram_channel(scheduled_post_id: int):
+        message=post.text_for_posting,
+        attachments=[],
+    )
+    if vk_response.status_code != requests.codes.ok or 'error' in vk_response.json():
```
Но тогда надо подумать где делать вот эту проверку, наверное тоже в vkontakte.py

vladimirshkoda replied:
Вынес эту проверку внутрь VkAPI клиента.

vladimirshkoda committed:
Add doc attachment support

vladimirshkoda committed:
Refactor Vk API requests.

Collect all Vk API requests into a small client.

b0g3r commented on postpost/api/vkontakte.py:
```diff
@@ -0,0 +1,115 @@
+        super(VkApiError, self).__init__(message)
+
+
+class VkApi(object):
```
Мне не нравится идея делать класс ради группировки группы методов, может сделаем плоским модуль? Или зачем такое

b0g3r replied:
В разговоре услышал. Напиши в докстринг, что эти методы объеденены в класс, так как во всех них используется общий для запросов токен и всякое такое

b0g3r replied:
А переименовать возможно стоит `VkAPI`

vladimirshkoda replied:
Напишу в док-стринг и переименую.

b0g3r commented on postpost/api/vkontakte.py:
```diff
@@ -0,0 +1,115 @@
+    return vk_api.send_post_to_group_wall(group_id, message, attachments)
+
+
+class VkApiError(Exception):
```
Правило именования флейка настойчиво просит именовать аббревиатуры как аббревиатуры: `api` и `API`, но не `Api`

b0g3r replied:
Если как мы решили в разговоре делать с True/False+error (привет, go), то можно и без экспешена обойтись

b0g3r replied:
При этом внутри бы `send_post_to_group` я бы не писал изменения PlatformPost, я бы делал установку статуса и установку ошибки всё равно снаружи посылалки, т.е. что-то такое
```success, error_text = send_post_to_vk(platform_post)  # return Tuple[bool, Optional[str]]
if success:
  # set status success
else:
  # set status failed
  # set error text 
```
а всё важное логировать где-то внутри send_post_to_vk, наружу отдавать только чистый интерфейсик

vladimirshkoda replied:
Сейчас исправлю.

b0g3r replied:
Вот тут я писал, что наружу возвращать статус и ошибку, но этого не хватит :(

Я уверен, что следующим шагом мы захотим хранить айдишники полученных постов или ещё чего-нибудь и тогда универсально сделать уже не получится. Пока пусть останется как есть, мне кажется это не критично. На следующей итерации поймём как лучше сделать

vladimirshkoda committed:
Add upload photo support and refactor several things.

- Wrap posting function call with try-except block.
- Group uploading doc function into one.

vladimirshkoda committed:
Rename Vk API client.

b0g3r commented on postpost/api/vkontakte.py:
```diff
@@ -0,0 +1,127 @@
+        response = self._request('wall.post', payload=payload)
+        return response
+
+    def upload_doc(self, doc_path: str) -> str:
```
Я бы либо в названии метода, либо бы в докстринге явно указал, что это используется у нас для гифок

vladimirshkoda replied:
Окей

vladimirshkoda replied:
Хотя нет, не окей: там потом вылезет какая-нибудь необходимость прикрепить док или таблицу, или ещё что-нибудь.

b0g3r commented on postpost/api/tasks.py:
```diff
@@ -37,8 +38,29 @@ def send_post_to_telegram_channel(scheduled_post_id: int):
+    Celery task which tries to send post to vk and changes status of platform post.
+    """
+    post = PlatformPost.objects.select_related('publication').get(id=scheduled_post_id)
+    try:
```
В try-except рекомендую засовывать только ту строчку, которая приведет к эксепшену, иначе код внутри разрастается

vladimirshkoda replied:
Дак а как ты поменяешь статус поста в зависимости от того, запостился он или нет? Ещё одну переменную заводить?

b0g3r replied:
try-except-else?
А вообще я тут где-то в комментариях предлагал избавляться снаружи от эксепшенов :thinking:

b0g3r replied:
Ой, у меня там ещё появилось, пойду откоментирую

vladimirshkoda replied:
Пока всё равно не понятно, где эту логику будем хранить, оставляю так.

vladimirshkoda reviewed (commented):
Короче, на данной итерации оставляю всё как есть, потому что работает и потому что непонятно, как это обобщить с другими соцсетями.

b0g3r commented:
Мне кажется codacy должен починиться с следующим пушем

teners commented on postpost/api/vkontakte.py:
```diff
@@ -0,0 +1,127 @@
+        upload_url = self._request(
+            'photos.getWallUploadServer',
+            {'group_id': group_id},
+        )['upload_url']
```
А не упадёт как-нибудь тут с KeyError?

vladimirshkoda replied:
Упадёт ещё в self._request, если что-то не так. Если всё так, то ссылку должен вернуть

teners replied:
Fair enough

teners commented on postpost/api/vkontakte.py:
```diff
@@ -0,0 +1,127 @@
+            {'group_id': group_id},
+        )['upload_url']
+
+        response = requests.post(upload_url, files={'file': open(photo_path, 'rb')})
```
Вот тут (и где-то ещё где файлы открываешь), может, всё-таки, в переменную файл открывать и проверять что он на самом деле открылся? Ты OSError тут не ловишь

vladimirshkoda replied:
Ну хз, лучше вообще file-like object сюда передавать, как по мне, типа какой-нибудь IO.

teners replied:
Ну я не предлагаю оставить как есть и просто ловить OSError, я имею в виду, что написано как написано, и _даже_ не ловишь экспешон

vladimirshkoda replied:
Done

vladimirshkoda committed:
Change several signatures.

- Send post function takes whole post instead of parts.
- Vk API takes IO instead of file path.

teners reviewed (approved):
lgtm, for now 🤷‍♀️
